tiledb/api/src/filter.rs

```diff
@@ -15,11 +19,227 @@ pub enum CompressionType {
     Zstd,
 }
 
+#[derive(Copy, Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
 pub enum ChecksumType {
     Md5,
     Sha256,
 }
 
+#[derive(Copy, Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
+pub enum WebPFilterInputFormat {
+    None,
+    Rgb,
+    Bgr,
+    Rgba,
+    Bgra,
+}
+
+impl WebPFilterInputFormat {
+    pub(crate) fn capi_enum(&self) -> u32 {
+        let ffi_enum = match *self {
+            WebPFilterInputFormat::None => ffi::WebPFilterInputFormat::NONE,
+            WebPFilterInputFormat::Rgb => ffi::WebPFilterInputFormat::RGB,
+            WebPFilterInputFormat::Bgr => ffi::WebPFilterInputFormat::BGR,
+            WebPFilterInputFormat::Rgba => ffi::WebPFilterInputFormat::RGBA,
+            WebPFilterInputFormat::Bgra => ffi::WebPFilterInputFormat::BGRA,
+        };
+        ffi_enum as u32
+    }
+}
+
+impl TryFrom<u32> for WebPFilterInputFormat {
+    type Error = crate::error::Error;
+    fn try_from(value: u32) -> TileDBResult<WebPFilterInputFormat> {
+        match value {
+            0 => Ok(WebPFilterInputFormat::None),
+            1 => Ok(WebPFilterInputFormat::Rgb),
+            2 => Ok(WebPFilterInputFormat::Bgr),
+            3 => Ok(WebPFilterInputFormat::Rgba),
+            4 => Ok(WebPFilterInputFormat::Bgra),
+            _ => Err(Self::Error::from(format!(
+                "Invalid webp filter type: {}",
+                value
+            ))),
+        }
+    }
+}
+
+#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
+pub struct CompressionData {
+    pub kind: CompressionType,
+    pub level: Option<i32>,
+    pub reinterpret_datatype: Option<Datatype>,
+}
+
+impl CompressionData {
+    pub fn new(kind: CompressionType) -> Self {
+        CompressionData {
+            kind,
+            level: None,
+            reinterpret_datatype: None,
+        }
+    }
+}
+
+#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
+pub enum FilterData {
+    None,
+    BitShuffle,
+    ByteShuffle,
+    BitWidthReduction {
+        max_window: Option<u32>,
```

Comment:
Why is just this setting optional?

Reply:
That's a bit ambiguous in hindsight. I meant to ask, of all the options for all the filter types, why is just this max_window optional?

Reply:
The answer is mostly laziness, this is the only one that came up in the tests where I had to actually write down the value.  And of course I didn't know it, so I decided to go on to option type.

Really all of these or none of these should be option.